### ISSUE TITLE
Obligatory and optional dependencies

### ISSUE THREAD

**Issue by TheChymera** [mikedh/trimesh#590]
How can I tell which dependencies are actually needed? The dependency list is IMHO pretty long, and a few of the packages are not automatically available for my distribution. Interestingly, however, the test suite seems to pass even without glooey, triangle, python-fcl, pyinstrument, and coveralls.

I also excluded e.g. scikits-learn on purpose, and the test suite passes even without that. I notice in `setup.py` lists only setuptools and numpy as minimal requirements, but without Shapely I get:

```
________________________________ ERROR collecting tests/test_packing.py _________________________________
ImportError while importing test module '/var/tmp/portage/dev-python/trimesh-3.3.5/work/trimesh-3.3.5/tests/test_packing.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib64/python2.7/site-packages/_pytest/python.py:512: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
/usr/lib64/python2.7/site-packages/py/_path/local.py:668: in pyimport
    __import__(modname)
/usr/lib64/python2.7/site-packages/_pytest/assertion/rewrite.py:301: in load_module
    exec(co, mod.__dict__)
tests/test_packing.py:5: in <module>
    from shapely.geometry import Polygon
E   ImportError: No module named shapely.geometry
```

When trying to run the test suite... So which then is the *actual* list of non-optional dependencies?

**Comment by mikedh** [comment]
Hey, all tests will not pass with just the minimal requirements but large parts of the library (i.e. the primary use case of "load a mesh and look at some vertices and faces as numpy arrays, and maybe get some inertia matrices) will be fine.

We run the tests in `tests/basic.list` for the minimal install:
`cd tests; pytest  -p no:warnings `grep -v '^#' basic.list`; cd ..`

"All tests passing" requires the "all" requirements, which you can dump into a requirements.txt if necessary: 
```
mikedh@cue:trimesh$ python setup.py --list-all
pillow
psutil
python-fcl
scipy
scikit-image
chardet
pycollada
xxhash
requests
rtree
msgpack
lxml
networkx
colorlog
pyglet
shapely
sympy
triangle
setuptools
jsonschema
svg.path
glooey
```

**Comment by TheChymera** [comment]
@mikedh are there any particular modules I can also skip installing if I just want the basic functionality? It doesn't make much sense to provide files which will be broken given the dependency resolution.

**Comment by mikedh** [comment]
Hey, I think it depends a lot on what you need to do. I think most people are best off with `pip install trimesh[easy]`, but if you are just opening an STL the minimal install is fine.

**Comment by TheChymera** [comment]
@mikedh yes, I understand, but [`setup.py`](https://github.com/mikedh/trimesh/blob/14ae942bf7306e9eaab1fc6ffa74cf8ec7934ddf/setup.py#L99) seems to indicate that all “packages” (or as I would more unambiguously call them, submodules) get installed no matter what. I am wondering whether there is any correspondence between the dependency selection and the submodules which will actually work for a given dependency set.

Gentoo Linux extensively uses a feature [called USE flags](https://wiki.gentoo.org/wiki/Handbook:AMD64/Working/USE) where you can specify in the package manager (one package manager for all packages, no pip+npm+God-knows-what) what features you want to select. I would like to leverage the modularity you have built into your dependency list to make the distribution of your package on Gentoo more flexible.

**Comment by mikedh** [comment]
Ah gotcha, yeah every submodule is included regardless of the install requested, which turns missing dependencies (except numpy) into explicit run-time errors rather than install-time.  There are definitely pros and cons to this approach, I just didn't want to force everyone to install a million things to just get an STL loader. 

Unfortunately I've somewhat neglected install documentation, but here's a rundown of how I see the various dependencies as of today:

`minimal`: numpy
Basic loaders (PLY, STL, GLTF), transformations, inertia calculations, etc. 
`minimal+`: numpy, scipy, networkx, lxml
Convex hulls, most graph stuff, etc. Maybe 75% functional with these four?
`easy`
Do path stuff, and really anything from pypi that installs cleanly (with the exception of `shapely` and `rtree`, which still unfortunately require external libraries as of the last time I checked, but they are necessary for pretty much all of the 2D path stuff). Probably 95%+ functional with these.
`all` 
Includes stuff for rarely-used functions, like GLTF schema checking, marching cubes, polygon triangulation, etc.


Happy to add things that would help with the USE flags, it seems like a nice solution!

**Comment by TheChymera** [comment]
Sounds great, also happy to hear you would like to help. Can the modules somehow be divided with say modules a and b for minimal; a, b, and c for minimal+; a, b, c, and d for easy; and a, b, c, d, and e for all?
